bundles/org.eclipse.osgi/felix/src/org/apache/felix/resolver/util/ShadowList.java

```diff
@@ -64,13 +65,18 @@ public void insertHostedCapability(ResolveContext context, HostedCapability wrap
             m_original.remove(removeIdx);
             unmodifiable.remove(removeIdx);
         }
-        int insertIdx = context.insertHostedCapability(m_original, toInsertCapability);
+        Capability[] copy = m_original.toArray(new Capability[0]); // make a copy so we can modify it and the caller can
+                                                                   // not modify ours
+        for (int i = 0; i < copy.length; i++) {
+            Capability capability = copy[i];
+            if (capability instanceof ShadowedCapability) {
+                // we unwrap the ShadowedCapability here as we always must pass only what we
+                // have got from the ResolveContext#findProviders.
+                copy[i] = ((ShadowedCapability) capability).getShadowed();
+            }
+        }
+        int insertIdx = context.insertHostedCapability(Arrays.asList(copy), toInsertCapability);
```

Comment:
This seems odd and not in the spirit of the specification when multiple hosted capabilities are inserted into the list.  The `ResolveContext` will keep getting the same original list back and it will not have the past hosted capabilities there that the `ResolveContext` had inserted.  This means the `ResolveContext` will not have any control over the order of the inserted `HostedCapability` objects in relation to each other.

Reply:
I now fixed that, the main problem is that Felix resolver is violation the contract of insertHosteCapability, I [added a TCK that shows the problem](https://github.com/osgi/osgi/pull/690), but at least this should work as before.